### DISCUSSION TITLE
What is the purpose of the reset_fwrd vs reset_no_fwrd

### DISCUSSION THREAD

**Discussion by stephendpmurphy** [ARMmbed/DAPLink#802] [Q&A] [answered]
I'm looking to make a Dip Dap clone and was just browsing the HIC specific source for the LPC11U35 and saw some odd things in the source regarding the "forwarded reset".

It looks like both functions are defined (fwrd & no fwrd) and in the HDK both of these pins are hooked up and driven from the Reset PB on the board (Referencing the HDK repository schematics).

My question is: what's the practical use of having both of these defined in the source as inputs when the ```gpio_get_reset_btn(void)``` just checks both and ORs them? In connecting the hardware this way, doesn't this mean you would never actually hit the DAPLink bootloader (when depressing the RESET button at power up) and you would always boot into the built into the ROMbootloader of the LPC since PIO0_1 is the ROMboot ISP pin?

Follow-up question: Is that the intention? For the Dip Dap hardware to use the ROM bootloader and only run the DAPLink interface FW? 

### gpio.c
```c
uint8_t gpio_get_reset_btn_no_fwrd()
{
    return LPC_GPIO->PIN[PIN_RESET_IN_PORT] & PIN_RESET_IN ? 0 : 1;
}

uint8_t gpio_get_reset_btn_fwrd()
{
    return LPC_GPIO->PIN[PIN_RESET_IN_FWRD_PORT] & PIN_RESET_IN_FWRD ? 0 : 1;
}
```
### gpio.h
```c
static inline uint8_t gpio_get_reset_btn(void)
{
    return gpio_get_reset_btn_no_fwrd() || gpio_get_reset_btn_fwrd();
}
```

**Comment by mathias-arm** [comment, marked as answer]
DAPLink for LPC11U35 does not use a bootloader (there is no `lpc11u35_bl`).

AFAIK, there are 3 uses of the reset buttons:
- `gpio_get_reset_btn()` in `source/daplink/drag-n-drop/vfs_user.c` for file automation in both bootload and interface.
- `gpio_get_reset_btn()` in `source/daplink/bootloader/main_bootloader.c` to stop bootloader from starting interface.
- `gpio_get_reset_btn_fwrd()` in `source/daplink/interface/main_interface.c` to reset the target.

The bootloader does not need two buttons (both perform the same operations). With two buttons, the interface would have one (`gpio_get_reset_btn_fwrd()`) that reset the target and performs file automation and one that only does the latter (`gpio_get_reset_btn_no_fwrd()`). If the probe only has one button it would most likely be wired to the `gpio_get_reset_btn_fwrd()` function and the consequence is that you would not be able to run file automation without resetting the system as well.
